Sprint 11

s4fonovoleg opened the pull request.

s4fonovoleg committed:
sprint 11 task

s4fonovoleg committed:
Фикс иконки в tab bar
Цвет фона в профиле
Убрано удаление токена

LizaIlinykh commented on ImageFeed/Splash/SplashViewController.swift:
```diff
@@ -67,13 +94,42 @@ extension SplashViewController: AuthViewControllerDelegate {
 	}
+	
+	private func fetchProfile(token: String) {
+		profileService.fetchProfile(token) { result in
```
здесь может быть утечка памяти, так как completion у метода fetchProfile является сбегающим замыканием и при этом мы сильно захватываем self. сейчас утечки нет, так как замыкание нигде не сохраняется, но я советую все же всегда перестраховываться и делать ссылку слабой

LizaIlinykh commented on ImageFeed/Splash/SplashViewController.swift:
```diff
@@ -67,13 +94,42 @@ extension SplashViewController: AuthViewControllerDelegate {
+
+			switch result {
+			case .success(let profile):
+				ProfileImageService.shared.fetchProfileImageURL(username: profile.username) { _ in }
```
часто completion для методов делают опциональным, это поможет избавиться от бесполезных хвостиков. 
имею в виду этот: { _ in }

s4fonovoleg replied:
но тогда придется передавать здесь вторым параметром nil, если я правильно понимаю

LizaIlinykh replied:
можно задать значение по умолчанию и тогда параметр можно совсем не указывать
func fetchProfileImageURL(username: String, _ completion: ((Result<String, Error>) -> Void)? = nil)

LizaIlinykh commented on ImageFeed/Auth/OAuth2Service.swift:
```diff
@@ -26,6 +31,12 @@ final class OAuth2Service {
 	///   - completion: метод обработки результата.
 	func fetchAuthToken(code: String, completion: @escaping (Result<String, Error>) -> Void) {
+		assert(Thread.isMainThread)
+		if lastCode == code { return }
```
здесь идеально подойдет guard вместо if =)

LizaIlinykh commented on ImageFeed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,86 @@
+final class ProfileImageService {
+	static let DidChangeNotification = Notification.Name(rawValue: "ProfileImageProviderDidChange")
+	
+	static let shared = ProfileImageService()
```
при создании синглтона стоит заприватить его инициализатор
https://medium.com/@alessandromanilii/singleton-pattern-in-swift-5c7acde96c3e

LizaIlinykh commented on ImageFeed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,86 @@
+	}
+}
+
+struct UserResult: Codable {
```
модели лучше помещать в отдельные файлы

LizaIlinykh commented on ImageFeed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,91 @@
+import Foundation
+
+final class ProfileService {
+	static let shared = ProfileService()
```
здесь тоже нужно заприватить инициализатор

LizaIlinykh commented on ImageFeed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,91 @@
+	}
+}
+
+struct ProfileResult: Codable {
```
эти модельки тоже можно вынести в отдельные файлы

LizaIlinykh commented on ImageFeed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,91 @@
+			let response = result.flatMap { data -> Result<ProfileResult, Error> in
+				Result { try decoder.decode(ProfileResult.self, from: data) }
+			}
+			completion(response)
```
действия c UI всегда должны выполняться на главном потоке, поэтому давай completion вынесем на main thread при помощи DispatchQueue.main.async

LizaIlinykh commented on ImageFeed/Splash/SplashViewController.swift:
```diff
@@ -67,13 +94,42 @@ extension SplashViewController: AuthViewControllerDelegate {
+				fetchProfile(token: token)
 			case .failure:
-				
+				UIBlockingProgressHUD.dismiss()
```
при ошибке во время запрос токена тоже необходимо показать алерт с ошибкой

s4fonovoleg committed:
Добавлены слабые ссылки
Модели вынесены в отдельные файлы
Выполнение completion вынесено на главный поток
Добавлен алерт об ошибке запроса токена
Опциональные замыкания
Guard вместо if

LizaIlinykh commented on ImageFeed/AlertHelper.swift:
```diff
@@ -0,0 +1,14 @@
+import UIKit
+
+class AlertHelper {
```
здесь было бы здорово вместо создания отдельного класса просто расширить класс UIViewController и тогда гораздо удобнее будет использовать =)

extension UIViewController {
	func showErrorAlert(title: String, message: String, actionTitle: String ) {
		let alert = UIAlertController(
			title: title,
			message: message,
			preferredStyle: .alert)
		let action = UIAlertAction(title: actionTitle, style: .default)

		alert.addAction(action)
		present(alert, animated: true)
	}
}

s4fonovoleg replied:
точно... мой уставший пятничный мозг не сообразил, спасибо!)

s4fonovoleg committed:
Алерт перенесен в расширение
